File: service/static/runtime/math.js
// JS runtime shims for math/system functions imported by wasm
// Keep names exactly as in MangledName fields. IO functions moved to io.js

export function sqrt(x) { return Math.sqrt(x); }

export function sin(x) { return Math.sin(x); }
export function cos(x) { return Math.cos(x); }
export function tan(x) { return Math.tan(x); }
export function cotan(x) { return 1.0 / Math.tan(x); }
export function asin(x) { return Math.asin(x); }
export function acos(x) { return Math.acos(x); }
export function atan(x) { return Math.atan(x); }
export function log(x) { return Math.log(x); }
export function log10(x) { return Math.log10 ? Math.log10(x) : Math.log(x) / Math.LN10; }
export function exp(x) { return Math.exp(x); }

export function div_qum(a, b) {
  const ai = BigInt(a), bi = BigInt(b);
  if (bi === 0n) return 0n;
  // Mathematical division: floor-style so that
  // a = div_qum(a,b) * b + mod_qum(a,b) with 0 <= mod < |b|.
  let q = ai / bi; // trunc towards zero for BigInt
  let r = ai % bi;
  if (r !== 0n && ((r > 0n) !== (bi > 0n))) {
    q -= 1n;
  }
  return q;
}
export function mod_qum(a, b) {
  const ai = BigInt(a), bi = BigInt(b);
  if (bi === 0n) return 0n;
  // Remainder consistent with mathematical division above:
  // 0 <= mod_qum(a,b) < |b| and a = div_qum(a,b) * b + mod_qum(a,b).
  let r = ai % bi;
  if (r !== 0n && ((r > 0n) !== (bi > 0n))) {
    r += bi;
  }
  return r;
}

export function fpow(a, n) {
  // a: number (double), n: int
  return Math.pow(a, Number(n));
}

export function pow(a, b) { return Math.pow(a, b); }

export function trunc_double(x) { return (x < 0 ? BigInt(Math.ceil(x)) : BigInt(Math.floor(x))); }

export function rand_double(x) { return Math.random() * x; }
export function rand_double_range(a, b) { return a + Math.random() * (b - a); }
export function rand_int64(x) { return BigInt(Math.floor(Math.random() * Number(x))); }
export function rand_int64_range(a, b) {
  const A = BigInt(a), B = BigInt(b);
  if (B <= A) return A;
  const span = B - A;
  const r = BigInt(Math.floor(Math.random() * Number(span)));
  return A + r;
}

// Common compiler-rt builtins that may be referenced when --allow-undefined is used.
// Minimal shims to satisfy linkage; adjust if exact semantics are needed.
export function __multi3(a, b) {
  // return lower 64 bits of 128-bit product
  const MOD64 = (1n << 64n) - 1n;
  return (BigInt(a) * BigInt(b)) & MOD64;
}

File: qumir/runtime/math.cpp
#include "math.h"

#include <stdlib.h>
#include <math.h>
#include <limits.h>
#include <iostream>

#include <limits>

double cotan(double x) {
    return 1.0 / tan(x);
}

int64_t trunc_double(double x) {
    return static_cast<int64_t>(x);
}

double rand_double(double x) {
    // random on [0,x]
    return static_cast<double>(rand()) / RAND_MAX * x;
}

double rand_double_range(double a, double b) {
    // random on [a,b]
    return a + static_cast<double>(rand()) / RAND_MAX * (b - a);
}

int64_t rand_int64(int64_t x) {
    // random on [0,x]
    if (x == 0) return 0;
    return static_cast<int64_t>(rand()) % x;
}

int64_t rand_int64_range(int64_t a, int64_t b) {
    // random on [a,b]
    if (b <= a) return a;
    return a + static_cast<int64_t>(rand()) % (b - a);
}

int64_t div_qum(int64_t a, int64_t b) {
    if (b == 0) {
        // division by zero
        return 0;
    }
    // Mathematical division: round toward -infinity so that
    // a = div_qum(a,b) * b + mod_qum(a,b) with 0 <= mod < |b|.
    int64_t q = a / b;
    int64_t r = a % b;
    if (r != 0 && ((r > 0) != (b > 0))) {
        q -= 1;
    }
    return q;
}

int64_t mod_qum(int64_t a, int64_t b) {
    if (b == 0) {
        // division by zero
        return 0;
    }
    // Remainder consistent with mathematical division above:
    // 0 <= mod_qum(a,b) < |b| and a = div_qum(a,b) * b + mod_qum(a,b).
    int64_t r = a % b;
    if (r != 0 && ((r > 0) != (b > 0))) {
        r += b;
    }
    return r;
}

double fpow(double a, int n) {
    if (n < 0) {
        if (a == 0.0) {
            // division by zero
            return NAN;
        }
        return 1.0 / fpow(a, -n);
    }
    double result = 1.0;
    while (n) {
        if (n & 1) {
            result *= a;
        }
        a *= a;
        n >>= 1;
    }
    return result;
}
